bugfix(port-forward): Correctly handle known errors

sxllwx opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #74551 #107203

#### Special notes for your reviewer:

For detailed fault location process, please refer to https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-1514794473

containerd PR: https://github.com/containerd/containerd/pull/8418

This PR is mainly to correctly handle `EPIPE` errors. So we rely on the error returned by `h.forwarder.PortForward` [undecorated or the underlay err can be read by `errors.Is(err, syscall.EPIPE)`].

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed(port-forward) correctly handle the TCP connection receiving RST. Prevent `kubectl port-forward` from receiving `broken pipe` error
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

bart0sh commented on pkg/kubelet/cri/streaming/portforward/httpstream.go:
```diff
@@ -252,11 +258,26 @@ func (h *httpStreamHandler) portForward(p *httpStreamPair) {
+		klog.V(5).InfoS("done invoking forwarder.PortForward but got EPIPE", "conn", h.conn, "request", p.requestID, "port", portString)
+		p.dataStream.Reset()
+		p.errorStream.Reset()
+		streamHasBeenReset = true
```
Would it make sense to wrap it into a named func and use it in both places - here and when you defer it?

bart0sh replied:
May be I'm missing something, but wouldn't defer work as expected if  you leave only `return` here? Why to call `Reset` explicitly?

sxllwx replied:
Thanks for the timely review --- I've fixed it, PTAL thx.

bart0sh replied:
After looking at this again I'm wondering would it be clearer without defer and an additional function? After all we'll just need 2 calls of `Close` at the end of this function and two calls of 'Reset' here.

sxllwx replied:
I think defer is needed. In addition to the end of the code block execution, it will also be executed after panic occurs in `h.forwarder.PortForward`.

bart0sh commented:
@sxllwx is is a user-visible issue? If it is, please provide a release note.

/triage accepted
/priority important-longterm
/assign

sxllwx commented:
> @sxllwx is is a user-visible issue? If it is, please provide a release note.
> 
> /triage accepted /priority important-longterm /assign

release-note  added, thx ~

bart0sh commented:
/retest

bart0sh commented on pkg/kubelet/cri/streaming/portforward/httpstream.go:
```diff
@@ -242,8 +243,10 @@ Loop:
-	defer p.errorStream.Close()
+	resetStreams := false
+	defer func() {
+		closeStreamPair(p, resetStreams)
```
Now this looks like `closeStreamPair` is not actually needed. Would it work if you just move its code here?

sxllwx replied:
fixed, PTAL

bart0sh replied:
This is the code you had previously. It looks ok to me, but I was thinking about getting rid of an extra function and keeping all reset/close logic in one place, something like this may be?
resetStreams = false
```
defer func() {
	if resetStreams {
		p.dataStream.Reset()
		p.errorStream.Reset()
		return
	}
	p.dataStream.Close()
	p.errorStream.Close()
}
```

sxllwx replied:
Sorry, I missed your point 😅. A fix has been made,


I remember you asked a question before [If you haven't asked, then here is a supplement]

> Why clear reset stream instead of close stream?

I tested the PR in Containerd [the main way is to debug spdy through `export DEBUG=1`] The strange thing is that close does not seem to close the entire stream. As a result, the client side will be blocked for a long time. After replacing it with reset, the problem will naturally disappear. Maybe it is a bug of spdy? Do you have an answer here?

aojea replied:
> As a result, the client side will be blocked for a long time. After replacing it with reset, the problem will naturally disappear. 

what does the client does when you reset the stream? does it exit clean?

aojea replied:
> Reset() closes both directions of the stream

aojea replied:
but why it hangs?
I don't have clear where this patch is needed @sxllwx , is this needed for containerd for k/k or for both?
I think we can add an ee2e test, is just to use a portforwrd to a pod with a listening server, stop the listening server and try to send traffic, then it will hit the EPIPE error because the forwarded stream will not have anything in the other side

sxllwx replied:
> does the client does when you reset the stream? does it exit clean?

yes~

sxllwx replied:
> but why it hangs? I don't have clear where this patch is needed @sxllwx , is this needed for containerd for k/k or for both? I think we can add an ee2e test, is just to use a portforwrd to a pod with a listening server, stop the listening server and try to send traffic, then it will hit the EPIPE error because the forwarded stream will not have anything in the other side

xref: https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-1543873482

aojea replied:
I've added an image to be able to add e2e tests https://github.com/kubernetes/kubernetes/pull/117442#issuecomment-1535532821 
and @brianpursley put together an e2e test https://github.com/kubernetes/kubernetes/pull/117851

can we respect the commits and aggregate everything in one PR?

bart0sh commented:
/retest

bart0sh commented:
/lgtm
/assign @SergeyKanzhelev @mrunalp 
/king bug

bart0sh commented:
/kind bug

liggitt commented:
xref https://github.com/kubernetes/kubernetes/pull/114854 , not sure if there were other fixes we should pull in

aojea commented on pkg/kubelet/cri/streaming/portforward/httpstream.go:
```diff
@@ -252,11 +261,24 @@ func (h *httpStreamHandler) portForward(p *httpStreamPair) {
+
+	// forwarding in one direction has been completed.
+	// other direction: we got RST, so we should reset data && err stream
+	if errors.Is(err, syscall.EPIPE) {
```
one question, if we got reset don't we get a `ECONNRESET` error?

aojea replied:
oh, I see, so the problem is when we write to a closed pipe, because the target container no longer runs anything

aojea replied:
the container not, but the process listening on that port, right?

sxllwx replied:
yes~

sxllwx replied:
> one question, if we got reset don't we get a ECONNRESET error?

xref: https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-1550655795

aojea replied:
yeah, but think that this code is in different places, so the same code in containerd and kubelet will handle different errors. If the connection has been closed by the RST and one side tries to write, it will get an EPIPE error.
I don't know how the `ECONNRESET` is surfaced though, when it tries to read I imagine

sxllwx replied:
ECONNRESET is another story, ECONNRESET is handled here because of brianpursley's suggestion.
xref: https://github.com/kubernetes/kubernetes/pull/117493#discussion_r1191478301

sxllwx commented:
> xref #114854 , not sure if there were other fixes we should pull in

I picked the #114854 stream-related changes into containerd for testing [codebase](https://github.com/sxllwx/containerd/tree/pick/114854), but it doesn't work so far.

aojea commented:
related https://github.com/kubernetes/kubernetes/pull/117442

we need e2e tests, this feature is so complex that is hard to understand the implications across the multiple components

https://github.com/kubernetes/kubernetes/blob/9924dc65b75975c5993da95ea8e0cb9360543c3f/test/e2e/kubectl/kubectl.go#L550-L565

aojea commented:
/cc @brianpursley

aojea commented:
@brianpursley @sxllwx in addition to the e2e test, something we REALLY NEED for this feature is to have integration tests for the portforward client-go/tools https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/portforward/portforward_test.go similar to the one that I added here https://github.com/kubernetes/kubernetes/pull/116191 , instead of mocking the dialer use a real httpserver and httpclient, so we can really exercise the network

Is this something one of you are up to take ?

sxllwx commented:
I will take on this task and add unit tests in addition to e2e.  I'll put them all in this PR.

brianpursley committed:
Add e2e test reproducing issue #111825

sxllwx commented on pkg/kubelet/cri/streaming/portforward/httpstream.go:
```diff
@@ -252,11 +261,24 @@ func (h *httpStreamHandler) portForward(p *httpStreamPair) {
+
+	// forwarding in one direction has been completed.
+	// other direction: we got RST, so we should reset data && err stream
+	if errors.Is(err, syscall.EPIPE) {
```
@aojea Thank you for your endorsement. I am adding a corresponding e2e test recently, can you help me add an image [or you can tell me the guidelines for adding a test image]?

sxllwx replied:
This repository stores the images I have prepared to be tested. 【I use golang in the image to compile to ensure our consistent experience. 】 https://github.com/sxllwx/k8s-74551-reproducer/releases/tag/v0.1.0

aojea replied:
but we already have an e2e test and an image that reproduces it https://github.com/kubernetes/kubernetes/pull/117493#discussion_r1191132124

sxllwx replied:
In fact, these are two different scenarios [although Rest happened in the end]:

#111825 occurs because the server RST the connection

#74551 It is because the http.Handler exited without reading the request.Body. From the pcap, we can see that the server sent a FIN to the client, but because the FIN-ACK has not been received for a long time, the server initiated an RST.

`registry.k8s.io/e2e-test-images/agnhost:2.44` does simulate the #111825 scenario, but fails to simulate the issue seen in #74551.

aojea replied:
> can you help me add an image [or you can tell me the guidelines for adding a test image]?

this is just an http server https://github.com/sxllwx/k8s-74551-reproducer/releases/tag/v0.1.0

`agnhost netexec` already provides that

sxllwx replied:
`agnhost netexec` is indeed fully reproducible. Turns out his source code is in test/images/agnhost/netexec.go. Great～

aojea replied:
lovely

sxllwx replied:
ahhhh, I have added the corresponding e2e test, please take a look?

brianpursley commented on pkg/kubelet/cri/streaming/portforward/httpstream.go:
```diff
@@ -252,11 +261,24 @@ func (h *httpStreamHandler) portForward(p *httpStreamPair) {
+	// forwarding in one direction has been completed.
+	// other direction: we got RST, so we should reset data && err stream
+	if errors.Is(err, syscall.EPIPE) || errors.Is(err, syscall.ECONNRESET) {
+		klog.V(5).InfoS("done invoking forwarder.PortForward but got EPIPE", "conn", h.conn, "request", p.requestID, "port", portString)
```
Can this klog message say which one it got: EPIPE or ECONNRESET?

sxllwx replied:
Thanks for the reminder, it has been added.

brianpursley commented:
Question: Does this PR require https://github.com/containerd/containerd/pull/8418 as a prerequisite before the e2e tests can pass?

When I was testing ECONNRESET the other day, I had to upgrade my cluster to use a new containerd built from https://github.com/containerd/containerd/pull/8418 in order for it to work.

sxllwx commented:
> Question: Does this PR require https://github.com/containerd/containerd/pull/8418 as a prerequisite before the e2e tests can pass?

Yes, https://github.com/containerd/containerd/pull/8418 is a prerequisite for these e2e tests we added. As @aojea said, the problem caused by this bug is very complicated, and I need to enrich the comments to try my best to explain everything that happened.

aojea commented on test/e2e/kubectl/portforward.go:
```diff
@@ -207,6 +241,45 @@ func runPortForward(ns, podName string, port int) *portForwardCommand {
+	}
+
+	ginkgo.By("Running 'kubectl port-forward'")
+	cmd := runPortForward(f.Namespace.Name, pod.Name, 8080)
```
be careful with this common ports because there can be conflicts

aojea replied:
anyway, not important right now

aojea commented on test/e2e/kubectl/portforward.go:
```diff
@@ -492,6 +565,10 @@ var _ = SIGDescribe("Kubectl Port forwarding", func() {
 		})
+
+		ginkgo.It("should support server never read request, and disconnects", func(ctx context.Context) {
+			doTestConnectionNeverReadRequest(ctx, "localhost", f)
```
we need to add a skip indicating the issue and the reason, something like
```
e2eskipper.Skipf("Ref: https://issues.k8s.io/74551 dependency on the container runtime")
```

aojea replied:
otherwise the test will be failing

sxllwx replied:
fixed, PTAL

aojea commented:
> My testing shows this is necessary to fix the "connection reset by peer" problem. I tested this both manually and with https://github.com/kubernetes/kubernetes/pull/117851 and the above change fixes it in those scenarios. I haven't had a change to test with postgres/psql yet, but am hopeful it would do the same there.

@sxllwx can you please incorporate this e2e test https://github.com/kubernetes/kubernetes/pull/117851/commits/1d1a39828b5d9d79f8710667d8b1612fba6c7e41 from @brianpursley in this PR keeping him as author? (feel free to reach out in slack if you need help on how to do it)

sxllwx commented:
xref: https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-1550655795

sxllwx commented:
> 

Ok, my next job will be:

- Add some notes to help understand,
- Squash my own commits into one.
- cherry-pick @brianpursley https://github.com/kubernetes/kubernetes/commit/1d1a39828b5d9d79f8710667d8b1612fba6c7e41

bart0sh commented:
@sxllwx since this is a bugfix, would it make sense to cover it with e2e tests?

aojea commented:
> @sxllwx since this is a bugfix, would it make sense to cover it with e2e tests?

yeah, there has to be 2, there is discussion on some comments, but one has to be skipped because it depends on containerd/crio to revendor this code

brianpursley commented on test/e2e/kubectl/portforward.go:
```diff
@@ -32,9 +32,13 @@ import (
 	"syscall"
 	"time"
 
+	"k8s.io/apimachinery/pkg/util/intstr"
```
nit: can you group these with the other imports below?

sxllwx replied:
thx for review, fixed, PTAL

brianpursley reviewed (changes requested):
Looks good.  Just one nit comment about import grouping.

I guess we need to figure out how to get some attention for https://github.com/containerd/containerd/pull/8418, and then once that is merged, circle back to un-skip these e2e tests.

sxllwx committed:
Add e2e test reproducing issue #74551

sxllwx commented:
> Looks good. Just one nit comment about import grouping.
> 
> I guess we need to figure out how to get some attention for [containerd/containerd#8418](https://github.com/containerd/containerd/pull/8418), and then once that is merged, circle back to un-skip these e2e tests.

I have watched this PR, when he is merged, I will adjust SKIP. thanks for reminding.

sxllwx commented:
/retest-required

aojea commented on test/e2e/kubectl/portforward.go:
```diff
@@ -492,6 +657,18 @@ var _ = SIGDescribe("Kubectl Port forwarding", func() {
+
+	ginkgo.Describe("with a server that sends RST upon accepting a connection", func() {
+		ginkgo.It("should connect, send data, and then connect again", func(ctx context.Context) {
+			e2eskipper.Skipf("Ref: https://issues.k8s.io/74551 dependency on the container runtime")
```
this test should work with your fix, we should not skip it

brianpursley replied:
I think the reason this is skipped is that it requires a new version of containerd with https://github.com/containerd/containerd/pull/8418, which has yet to be merged.

When I tested, I built and used a version of containerd with that fix *in addition to* the fix in this PR and *both* PRs were required in order for this e2e test to pass.

sxllwx replied:
> Perhaps we should clarify a bit here.

Our goal here is to deal with [EPIPE, ECONNRESET] reasonably, and these two related e2e tests need to rely on [Containerd's PR](https://github.com/containerd/containerd/pull/8418) to be merged.

aojea replied:
can you do a temporal commit to remove the skip here so we can see the result on the CI?
 I think that this should work without the containerd changes, but I may be wrong
I would test it myself but the CI is faster and I'm a bit busy :)

sxllwx replied:
PTAL

aojea replied:
ok, thanks for doing the exercise, appreciate it

sxllwx commented on test/e2e/kubectl/portforward.go:
```diff
@@ -255,6 +351,75 @@ func doTestConnectSendDisconnect(ctx context.Context, bindAddress string, f *fra
+
+	ginkgo.By("Running 'kubectl port-forward'")
+	func() {
+		cmd := runPortForward(f.Namespace.Name, pod.Name, 8080)
```
fixed, thx~

apjoseph commented:
@brianpursley@sxllwx I've been following this PR as I  have Postgres instances that use SSL and so kubectl port-forward is effectively unusable for me until this update is released.

When this fix is pushed, will it only require an update of the kubectl client or will it require a cluster upgrade as well?

Thanks for all your work on this critical bug!

brianpursley commented:
> When this fix is pushed, will it only require an update of the kubectl client or will it require a cluster upgrade as well?

Unfortunately, this would require an existing cluster to be upgraded.  It should not require a kubectl upgrade.

After this PR is merged, I'm hoping it can be be cherry-picked back to previous kubernetes versions that are still in the support window, which should make it a little bit easier to upgrade.

sxllwx commented:
@brianpursley Thank you for your efforts. 😄

> Is anything required to update the e2e tests to use the new containerd version(s)?

The skipped test cases in e2e can run correctly only if our e2e environment needs to use the version containing https://github.com/containerd/containerd/pull/8418 containerd.

apjoseph commented:
@brianpursley @sxllwx Any idea or ETA on when this will be merged and rolled out? No pressure -just curious on the timing/process. 

I'm working on introducing Kubernetes at my organization and have serval internal users that need to connect to Postgres instances I'm trying to decide whether to wait for this fix or whether to build an interim auth solution. So any insight into the process would be very helpful!

sxllwx commented:
/lgtm

sxllwx commented:
renew:

In Containerd's [main branch](https://github.com/containerd/containerd/tree/33e544e94a67e3ac5c261101885773e2593cc923), the bugfix of https://github.com/containerd/containerd/pull/8418 was **rolled back**. The reason is that https://github.com/containerd/containerd/pull/9254 was merged

> After this PR is merged and Containerd updates Staging, the problems mentioned in https://github.com/kubernetes/kubernetes/issues/74551 in the Containerd main branch can be fixed.

Fortunately, the latest release of containerd does not use https://github.com/containerd/containerd/pull/9254

sxllwx commented:
One more thing. Currently, both cri-o and containerd tend to (or are already using) vendoring the streaming package rather than forking it.

sxllwx commented:
/retest-required

sxllwx commented:
/retest-required

sxllwx commented:
The e2e test cases added by this PR have been run in our test-infra.

Currently, our test-infra uses the `v1.7.13` version of containerd.

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/117493/pull-kubernetes-e2e-kind/1760956453159440384/artifacts/kind-worker2/containerd.log

/ping @bart0sh @aojea @liggitt @SergeyKanzhelev

aojea commented:
I fail to understand the third commit, sorry, there are a bunch of changes I can't understand, it feel ssome of them can also incorporated on the previous commits

sxllwx committed:
bugfix(port-forward): Correctly handle known errors

These two errors can occur in the following scenarios:

ECONNRESET: the target process reset connection between CRI and itself.
see: <https://github.com/kubernetes/kubernetes/issues/111825> for detail

EPIPE: the target process did not read the received data, causing the
buffer in the kernel to be full, resulting in the occurrence of Zero Window,
then closing the connection (FIN, RESET)
see: <https://github.com/kubernetes/kubernetes/issues/74551> for detail

In both cases, we should RESET the httpStream.

sxllwx commented:
> I fail to understand the third commit, sorry, there are a bunch of changes I can't understand, it feel ssome of them can also incorporated on the previous commits

I have sorted out the Commit, PTAL thx...

sxllwx commented:
/retest-required

aojea commented:
@sxllwx thanks for your patience, we are going to prioritize https://github.com/kubernetes/kubernetes/pull/123413 since code freeze is in 2 weeks and is an important new feature that users are waiting for long time and it is more complex to change than this PR.  once that merges we'll get this fix in

aojea commented:
to be clear, the goal is to merge this PR this release, is just an organizational issue

sxllwx commented:
> to be clear, the goal is to merge this PR this release, is just an organizational issue

Thank you for your guidance and work, and follow the rhythm of the organization. 😄

aojea commented:
/lgtm
/assign @liggitt

sxllwx commented:
/ping @liggitt PTAL thx~

sxllwx commented:
/ping @liggitt PTAL thx~

eddiezane commented:
This is a really needed fix for what we've been chasing down in kubectl port-forward for a few years now. At this point I don't think it will make 1.30.0 but hopefully we can get final approval and cherry pick it to 1.30.1, 1.29, and 1.28.

It also fixes several other issues of community projects that utilize port forward like https://github.com/defenseunicorns/zarf/issues/2104.

apjoseph commented:
@aojea @liggitt Is there anything blocking this from being merged and released?  The bug that this pull request is intended to fix is a pretty catastrophic longstanding issue -since it breaks TLS connections over kubectl which affects vast numbers of services and development workflows AND most importantly there is no (known/documented) workaround other than disabling TLS which is not always an option.

liggitt commented:
> is there anything blocking this from being merged and released

just reviewer time to understand the fix and whether there it is strictly more correct / has any possible side effects that would break existing flows :-/

liggitt commented on staging/src/k8s.io/kubelet/pkg/cri/streaming/portforward/httpstream.go:
```diff
@@ -252,11 +261,34 @@ func (h *httpStreamHandler) portForward(p *httpStreamPair) {
+	}
+
+	if errors.Is(err, syscall.EPIPE) || errors.Is(err, syscall.ECONNRESET) {
+		// In the process of forwarding, we encountered error types that can be handled:
```
What does "can be handled" mean, exactly?

it looks like this change avoids telling the client an error occurred in this case... am I reading that right, and is that the intent? it's not clear why masking the existence of an error to the client is good or desired.

Doesn't handling it this way mean something on the client side would not be informed there was an error? Can't that cause problems on the client side?

liggitt replied:
Separately from whether we call Reset() or Close(), it's not clear to me why we don't write this error to `errorStream`

liggitt replied:
This is still outstanding... seems like we should try to tell the client there was an error. The questions I had on the tests triggering error conditions but not seeing error responses is outstanding as well

liggitt replied:
I think I understand what you are trying to do now. Currently, if we write *any* error back to the errorStream (which is per-subrequest), client-go currently reacts by tearing down the entire portforward connection here:

https://github.com/kubernetes/kubernetes/blob/d831b13e6f6fb5efb566100286190fedca6dd340/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L410-L415

If we want specific errors on specific subrequests to allow the overall portforward operation to continue, I think *that* is the place to tweak, but we should still write the error to the errorStream here

liggitt commented on staging/src/k8s.io/kubelet/pkg/cri/streaming/portforward/httpstream.go:
```diff
@@ -242,8 +243,16 @@ Loop:
+	resetStreams := false
+	defer func() {
+		if resetStreams {
+			p.dataStream.Reset()  //nolint: errcheck
+			p.errorStream.Reset() //nolint: errcheck
+			return
```
What does calling Reset() do, compared to calling Close()?

I had to page back in a little of the Reset() implementation, at first I thought this was trying to leave streams in a state where they could be retried or resumed. Is the intent to more completely / aggressively / immediately tear down and close the dataStream / errorStream?

sxllwx replied:
great question!

I have some explanations here, if you still have questions, please let me know. 😄

https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-1543873482

liggitt replied:
I do have a bunch of questions 😄 Happy to leave them here or meet to talk through them.

1. My understanding is that calling Close() is leaving things in a hung state, and the desire here by calling Reset() is to tear the connection down more aggressively / completely. Is that correct?

2. Is the Close() call itself blocking, or just leaving the stream or connection in a hung state?

3. Is there any benefit to calling Close() _and_ Reset() in this case? The docs inside resetStream appear to support this:

   ```go
   func (s *Stream) resetStream() error {
   	// Always call closeRemoteChannels, even if s.finished is already true.
   	// This makes it so that stream.Close() followed by stream.Reset() allows
   	// stream.Read() to unblock.
   ```

4. Both Close() and Reset() attempt to write data to the connection.

   Close():
   ```go
   	return s.WriteData([]byte{}, true)
   ```
   
   Reset() → resetStream():
   ```go
   	return s.conn.framer.WriteFrame(resetFrame)
   ```

   Is that write blocking? If so, should we set deadlines on the stream before calling Close() / Reset()?

sxllwx replied:
> Your question is great, I need to do some exploring.

I made some new discoveries while doing testing today, and I hope I can give some clues and conclusions tomorrow.

sxllwx replied:
The clues I have now are as follows (please correct me if I'm wrong):

kubectl port-forward will complete the establishment of a TCP connection with the remote end [kubectl ⇒ kube-apiserver ⇒ kubelet ⇒ containerd] the moment it starts ForwardPorts. Subsequent Connections received from the local Socket will be reused on this connection [just create **Stream** instead of **Conn**]

- https://github.com/kubernetes/kubernetes/blob/a9f04103854893056237a09250ad3335867b0391/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L345
- https://github.com/kubernetes/kubernetes/blob/a9f04103854893056237a09250ad3335867b0391/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L194

sxllwx replied:
As shown in the figure, 8080 (server) has started the FIN process with containerd (client), so io.EOF will be received at [io.Copy(stream, conn)](https://github.com/containerd/containerd/blob/5f440cbb79f60e7cda1978a83837771ba0a1574b/internal/cri/server/sandbox_portforward_linux.go#L86), but it is masked as nil by io.Copy

<img width="1874" alt="pcap" src="https://github.com/kubernetes/kubernetes/assets/11829907/cb390325-baa1-48e4-8cc4-1b53d3053eea">

sxllwx replied:
> My previous troubleshooting may have misled you. 😅

In fact, httpStream#Close() or httpStream#Reset() will not affect the **Coon**. The **Conn** needs Conn#Close() to close and unblock https://github.com/containerd /containerd/blob/03d81f595f4e0c1323591ea48275e96830915502/vendor/k8s.io/kubelet/pkg/cri/streaming/portforward/httpstream.go#L218

sxllwx replied:
Regarding your questions:

> My understanding is that calling Close() is leaving things in a hung state, and the desire here by calling Reset() is to tear the connection down more aggressively / completely. Is that correct?

httpStream#Close() will send Fin to the peer, but the local Stream cleaning work depends on the peer's reply to Fin (https://github.com/containerd/containerd/blob/e634f04d8cdb6c2f96eea2e4e66d0e4500a46282/vendor/github.com/moby/spdystream/ stream.go#L173)

> Is the Close() call itself blocking, or just leaving the stream or connection in a hung state?
     
httpStream#Close() will not block, but will put the stream in a huang state. After receiving the peer sending Fin from the httpStream, the local Stream will be cleaned up.

- https://github.com/containerd/containerd/blob/e634f04d8cdb6c2f96eea2e4e66d0e4500a46282/vendor/github.com/moby/spdystream/stream.go#L173


> Is there any benefit to calling Close() *and* Reset() in this case? The docs inside resetStream appear to support this: Is that write blocking? If so, should we set deadlines on the stream before calling Close() /Reset()?

SetDeadlines is meaningful, whether it is httpStream#Close() or httpStream#Reset(), the data packet will be sent.

> Both Close() and Reset() attempt to write data to the connection.Is that write blocking? If so, should we set deadlines on the stream before calling Close() / Reset()?

Neither httpStream#Close() nor httpStream#Reset() will block

sxllwx replied:
I hope I can answer your questions. If you still have questions, please let me know. 😄

sxllwx replied:
@liggitt PTAL ~ thx

liggitt replied:
>> Is there any benefit to calling Close() and Reset() in this case? The docs inside resetStream appear to support this:
>
> I think httpStream#Reset() is enough

Is Close() more graceful than Reset()? Is there a benefit to attempting the graceful close first, and then following it with the hard close of the stream?

sxllwx replied:
httpStream#Close() will send FIN to the peer. [You need to rely on the peer to send FIN (that is, the client calls httpStream#Close()) to clear the local stream] ref: https://www.chromium.org/spdy/spdy-protocol/spdy-protocol-draft3-1/#236-stream-half-close

httpStream#Reset() will remove the local stream and send RST

Because this is two-way traffic, the approach based on httpStream#Close() may be more graceful, but we need to decide how long the interval between Close and Reset is.

liggitt commented on test/e2e/kubectl/portforward.go:
```diff
@@ -207,6 +264,51 @@ func runPortForward(ns, podName string, port int) *portForwardCommand {
+		url := fmt.Sprintf("http://localhost:%d/header", cmd.port)
+		_, err := post(url, strings.NewReader(strings.Repeat("x", size)), nil)
+		if err != nil {
+			framework.Failf("post data to Pod: %v", err)
```
I'm a little confused here... this test ensures the post completes without error, but ... isn't it posting to a pod that never reads the input? wouldn't we *expect* that to error, rather than silently completing successfully?

liggitt commented on test/e2e/kubectl/portforward.go:
```diff
@@ -255,6 +357,75 @@ func doTestConnectSendDisconnect(ctx context.Context, bindAddress string, f *fra
+
+			ginkgo.By("Sending data to connection's local port")
+			if _, err := conn.Write([]byte("test\n")); err != nil {
+				framework.Failf("Error writing data to the server: %v", err)
```
Similar question here... it really looks like the change is making things report success to clients when they didn't actually succeed. Am I misreading that?

aojea replied:
I think the key of this problem is the UX people is expecting, portforward opens a proxy and forward TCP connections. Errors on TCP connections are signaled between endpoints in forms of RST or FIN packet to indicate the connection is finished, IIUIC in this case these errors are propagated as "closing the proxy", and people wants "close the connections but keep the proxy alive"

sxllwx replied:
After my testing, it seems to always be successful (no error will be returned). @brianpursley

brianpursley replied:
Correct, that line is unaffected by this PR, it will pass with the master branch, and with this PR.

The only part of the test that fails in master currently is the `"Attempting connection #2"` part.  This is what happens if you run this e2e test without this PR's changes:
```
  STEP: Attempting connection #1 @ 04/29/24 11:15:18.48
  STEP: Sending data to connection's local port @ 04/29/24 11:15:18.48
  STEP: Closing the connection to the local port @ 04/29/24 11:15:18.48
  STEP: Verifying logs @ 04/29/24 11:15:18.48
  STEP: Attempting connection #2 @ 04/29/24 11:15:29.529
  STEP: Connection couldn't connect to port @ 04/29/24 11:15:29.53
```

aojea commented:
> /lgtm /assign @liggitt

/lgtm cancel

Please check Jordan's comment, we commented offline, each port forward open two stream: data and error.

There are some errors we still need to log in the client but only close the corresponding streams and subsequently its existing TCP connections, but we can not swallow the errors in the middle, they need to be sent through the error stream

sxllwx commented:
Thanks for the review, we really should let the client know what happened to the server (what errors were encountered).

In order to better handle the errors we know, we may need a clear structure for information exchange (metav1.Status may not be very suitable for our scenario). I suggest using the following structure (we can consider using JSON for Codec) but now I don't know where to put it. Do you have any better suggestions?

typeError struct{
    Continue bool // Whether the service should continue. (Decided by the server)
    Message string
}

sxllwx committed:
bugfix(port-forward): Correctly handle known errors

- structured error types added
- use Close instead of Reset close data and error conn

sxllwx committed:
use unified method to handle errors encountered during portforward

sxllwx committed:
fix client-side and add unit tests

sxllwx committed:
Add comments to the portForwardErrResponse

sxllwx commented:
/test pull-kubernetes-unit

sxllwx commented:
PTAL @liggitt

dhirschfeld commented:
Ugh, a stale bot. 

/remove-lifecycle stale

aojea commented on staging/src/k8s.io/kubelet/pkg/cri/streaming/portforward/httpstream.go:
```diff
@@ -242,8 +242,8 @@ Loop:
 	ctx := context.Background()
-	defer p.dataStream.Close()
-	defer p.errorStream.Close()
+	defer p.dataStream.Close()  //nolint: errcheck
```
I wonder if we have the same problem as in here and we should Reset() instead of  Close()

https://github.com/kubernetes/kubernetes/pull/126718#discussion_r1782263866

aojea replied:
I see , discussed here https://github.com/kubernetes/kubernetes/pull/117493#discussion_r1543998504

soltysh commented:
A note for everyone who might be looking into this PR and trying it out with latest kind. [kind is using internally containerd](https://github.com/kubernetes-sigs/kind/blob/10e058c534cf6b8546be70f1eea2e414a08ff770/images/base/Dockerfile#L125-L126). Current version used there has the fix from https://github.com/containerd/containerd/pull/8418 included, which means you won't get failures from e2e-s in this PR. 

Newer versions of containerd, on the other hand [switched to relying on k8s provided streaming implementation](https://github.com/containerd/containerd/pull/9254), this will be available in 1.8 and 2.0 versions. So to be able to get the error back, and thus verify the functionality of this PR, one needs to [re-built the base image](https://kind.sigs.k8s.io/docs/contributing/development#building-the-base-image) with newer containerd and use that when building a node-image from this PR (`kind build node-image --base-image=...`).

liggitt commented:
synced offline with @aojea and @soltysh 

I still don't think plumbing up specific error details from the kubelet / containerd and skipping alerting the client that there was an error is the correct approach.

As I mentioned in https://github.com/kubernetes/kubernetes/pull/117493#discussion_r1568847701, the loop in the client that currently unconditionally tears down the overall streamConn when an error is seen handling a single portforward request seems like the place we should be making changes.

```diff
git diff
diff --git a/staging/src/k8s.io/client-go/tools/portforward/portforward.go b/staging/src/k8s.io/client-go/tools/portforward/portforward.go
index 83ef3e929b3..365b7dd1603 100644
--- a/staging/src/k8s.io/client-go/tools/portforward/portforward.go
+++ b/staging/src/k8s.io/client-go/tools/portforward/portforward.go
@@ -407,12 +407,22 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
        case <-localError:
        }
 
+       // This is from https://github.com/kubernetes/kubernetes/pull/126718
+       /*
+               reset dataStream to discard any unsent data, preventing port forwarding from being blocked.
+               we must reset dataStream before waiting on errorChan, otherwise, the blocking data will affect errorStream and cause <-errorChan to block indefinitely.
+       */
+       _ = dataStream.Reset()
+
        // always expect something on errorChan (it may be nil)
        err = <-errorChan
        if err != nil {
                runtime.HandleError(err)
-               pf.streamConn.Close()
+               // Don't tear down the whole parent port-forward pf.streamConn when there's an error handling a single request
+               // TODO: *something* has to notice and tear down the whole parent port-forward pf.streamConn when the backend is completely gone... what / where?
        }
+       // This also forces drain of the errorStream, similar to dataStream above
+       _ = errorStream.Reset()
 }
 
 // Close stops all listeners of PortForwarder.
```

liggitt commented:
@soltysh was going to dig into https://github.com/kubernetes/kubernetes/pull/117493#issuecomment-2414380075 more as well
